esp-hal-common/src/interrupt/riscv.rs

```diff
@@ -716,15 +503,307 @@ pub fn _setup_interrupts() {
+    const PLIC_MXINT_TYPE_REG: u32 = DR_REG_PLIC_MX_BASE + 0x4;
+    const PLIC_MXINT_CLEAR_REG: u32 = DR_REG_PLIC_MX_BASE + 0x8;
+    const PLIC_MXINT0_PRI_REG: u32 = DR_REG_PLIC_MX_BASE + 0x10;
+
+    /// Enable a CPU interrupt
+    pub unsafe fn enable_cpu_interrupt(which: CpuInterrupt) {
+        let cpu_interrupt_number = which as isize;
+        let mxint_enable = PLIC_MXINT_ENABLE_REG as *mut u32;
+        unsafe {
+            mxint_enable.write_volatile(mxint_enable.read_volatile() | 1 << cpu_interrupt_number);
+        }
+    }
+
+    /// Disable the given peripheral interrupt.
+    pub fn disable(_core: Cpu, interrupt: Interrupt) {
+        unsafe {
+            let interrupt_number = interrupt as isize;
+            let intr = &*crate::peripherals::INTERRUPT_CORE0::PTR;
+            let intr_map_base = intr.wifi_mac_intr_map.as_ptr();
+
+            // 0 doesn't disable the pripheral interrupt apparently but will raise a level 1
+            // interrupt - we use CPU interrupt 31 which is disabled after reset by default
+            intr_map_base.offset(interrupt_number).write_volatile(31);
+        }
+    }
+
+    /// Set the interrupt kind (i.e. level or edge) of an CPU interrupt
+    ///
+    /// This is safe to call when the `vectored` feature is enabled. The
+    /// vectored interrupt handler will take care of clearing edge interrupt
+    /// bits.
+    pub fn set_kind(_core: Cpu, which: CpuInterrupt, kind: InterruptKind) {
+        unsafe {
+            let intr = PLIC_MXINT_TYPE_REG as *mut u32;
+            let cpu_interrupt_number = which as isize;
+
+            let interrupt_type = match kind {
+                InterruptKind::Level => 0,
+                InterruptKind::Edge => 1,
+            };
+            intr.write_volatile(
+                intr.read_volatile() & !(1 << cpu_interrupt_number)
+                    | (interrupt_type << cpu_interrupt_number),
+            );
+        }
+    }
+
+    /// Set the priority level of an CPU interrupt
+    ///
+    /// Great care must be taken when using the `vectored` feature (enabled by
+    /// default). Avoid changing the priority of interrupts 1 - 15 when
+    /// interrupt vectoring is enabled.
+    pub unsafe fn set_priority(_core: Cpu, which: CpuInterrupt, priority: Priority) {
+        let plic_mxint_pri_ptr = PLIC_MXINT0_PRI_REG as *mut u32;
+
+        let cpu_interrupt_number = which as isize;
+        plic_mxint_pri_ptr
+            .offset(cpu_interrupt_number)
+            .write_volatile(priority as u32);
+    }
+
+    /// Clear a CPU interrupt
+    #[inline]
+    pub fn clear(_core: Cpu, which: CpuInterrupt) {
+        unsafe {
+            let cpu_interrupt_number = which as isize;
+            let intr = PLIC_MXINT_CLEAR_REG as *mut u32;
+            intr.write_volatile(1 << cpu_interrupt_number);
+        }
+    }
+
+    /// Get status of peripheral interrupts
+    #[inline]
+    pub fn get_status(_core: Cpu) -> u128 {
```

Comment:
We should probably move these out of the plic/classic modules as they are the same, also future PLIC targets may have less interrupts so really this is chip specific rather than interrupt module specific.

Reply:
🤔it's a good point - I think I will introduce a new cfg for that ... makes more sense than having it depend on the specific chip